File: backend/src/main/java/com/agentworkbench/harness/core/MessageHistoryNormalizer.java
package com.agentworkbench.harness.core;

import com.agentworkbench.harness.llm.ChatRequest;
import com.agentworkbench.session.entity.Message;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Ensures tool messages immediately follow the assistant message that issued their tool_calls.
 * <p>
 * During parallel tool execution, tool results may be persisted before the assistant message,
 * which breaks OpenAI message ordering when history is reloaded from DB.
 */
@Slf4j
public final class MessageHistoryNormalizer {

    private MessageHistoryNormalizer() {
    }

    public static List<Message> normalizeEntities(List<Message> messages, ObjectMapper objectMapper) {
        if (messages == null || messages.size() < 2) {
            return messages;
        }

        Map<String, Message> deferredTools = collectDeferredToolMessages(messages);
        if (deferredTools.isEmpty()) {
            return messages;
        }

        List<Message> normalized = new ArrayList<>(messages.size());
        for (Message msg : messages) {
            if ("TOOL".equals(msg.getRole())) {
                continue;
            }
            normalized.add(msg);
            if ("ASSISTANT".equals(msg.getRole()) && msg.getToolCalls() != null && !msg.getToolCalls().isEmpty()) {
                appendMatchingToolMessages(normalized, deferredTools, extractToolCallIds(msg.getToolCalls(), objectMapper));
            }
        }

        if (!deferredTools.isEmpty()) {
            log.warn("Dropping {} orphaned tool messages without a preceding assistant tool_calls",
                    deferredTools.size());
        }
        return normalized;
    }

    public static List<ChatRequest.Message> normalizeChatMessages(List<ChatRequest.Message> messages) {
        if (messages == null || messages.size() < 2) {
            return messages;
        }

        Map<String, ChatRequest.Message> deferredTools = new LinkedHashMap<>();
        for (ChatRequest.Message msg : messages) {
            if ("tool".equals(msg.getRole()) && msg.getToolCallId() != null) {
                deferredTools.put(msg.getToolCallId(), msg);
            }
        }
        if (deferredTools.isEmpty()) {
            return messages;
        }

        List<ChatRequest.Message> normalized = new ArrayList<>(messages.size());
        for (ChatRequest.Message msg : messages) {
            if ("tool".equals(msg.getRole())) {
                continue;
            }
            normalized.add(msg);
            if ("assistant".equals(msg.getRole()) && msg.getToolCalls() != null && !msg.getToolCalls().isEmpty()) {
                for (ChatRequest.ToolCall toolCall : msg.getToolCalls()) {
                    if (toolCall.getId() == null) {
                        continue;
                    }
                    ChatRequest.Message toolMsg = deferredTools.remove(toolCall.getId());
                    if (toolMsg != null) {
                        normalized.add(toolMsg);
                    }
                }
            }
        }

        if (!deferredTools.isEmpty()) {
            log.warn("Dropping {} orphaned tool messages without a preceding assistant tool_calls",
                    deferredTools.size());
        }
        return normalized;
    }

    private static Map<String, Message> collectDeferredToolMessages(List<Message> messages) {
        Map<String, Message> deferredTools = new LinkedHashMap<>();
        for (Message msg : messages) {
            if ("TOOL".equals(msg.getRole()) && msg.getToolCallId() != null) {
                deferredTools.put(msg.getToolCallId(), msg);
            }
        }
        return deferredTools;
    }

    private static void appendMatchingToolMessages(List<Message> normalized,
                                                    Map<String, Message> deferredTools,
                                                    List<String> toolCallIds) {
        for (String toolCallId : toolCallIds) {
            Message toolMsg = deferredTools.remove(toolCallId);
            if (toolMsg != null) {
                normalized.add(toolMsg);
            }
        }
    }

    private static List<String> extractToolCallIds(String toolCallsJson, ObjectMapper objectMapper) {
        List<String> ids = new ArrayList<>();
        try {
            List<ChatRequest.ToolCall> toolCalls = objectMapper.readValue(
                    toolCallsJson, new TypeReference<List<ChatRequest.ToolCall>>() {});
            for (ChatRequest.ToolCall toolCall : toolCalls) {
                if (toolCall.getId() != null) {
                    ids.add(toolCall.getId());
                }
            }
        } catch (Exception e) {
            log.warn("Failed to parse tool_calls while normalizing message history: {}", e.getMessage());
        }
        return ids;
    }
}

File: backend-ts/src/harness/core/message-history-normalizer.ts
import type { ChatMessage, ToolCall } from '../llm/chat-request.js';
import type { Message } from '../deps.js';
import { harnessLog } from '../log.js';

/** 发给模型的占位 tool 结果：补齐缺失的 tool_call 配对，避免网关 400 把会话卡死。 */
export const MISSING_TOOL_RESULT_PLACEHOLDER =
  '[系统] 该工具调用没有对应输出（执行中断或结果未保存）。请勿假定已成功；如仍需要请重试。';

export const MessageHistoryNormalizer = {
  normalizeEntities(messages: Message[] | null | undefined, parseToolCalls: (json: string) => ToolCall[]): Message[] | null | undefined {
    if (messages == null || messages.length < 2) return messages;

    const deferredTools = collectDeferredToolMessages(messages);
    if (deferredTools.size === 0) return messages;

    const normalized: Message[] = [];
    for (const msg of messages) {
      if (msg.role === 'TOOL') continue;
      normalized.push(msg);
      if (msg.role === 'ASSISTANT' && msg.toolCalls) {
        appendMatchingToolMessages(normalized, deferredTools, extractToolCallIds(msg.toolCalls, parseToolCalls));
      }
    }
    if (deferredTools.size > 0) {
      harnessLog('warn', `Dropping ${deferredTools.size} orphaned tool messages without a preceding assistant tool_calls`);
    }
    return normalized;
  },

  ensureContentPresent(messages: ChatMessage[] | null | undefined): void {
    if (messages == null) return;
    for (const msg of messages) {
      if (msg != null && msg.content == null) {
        msg.content = '';
      }
    }
  },

  normalizeChatMessages(messages: ChatMessage[] | null | undefined): ChatMessage[] | null | undefined {
    if (messages == null || messages.length === 0) return messages;
    if (!hasToolCallsOrToolMessages(messages)) return messages;

    const deferredTools = new Map<string, ChatMessage>();
    for (const msg of messages) {
      if (msg.role === 'tool' && msg.toolCallId != null && msg.toolCallId !== '') {
        deferredTools.set(msg.toolCallId, msg);
      }
    }

    const normalized: ChatMessage[] = [];
    let filled = 0;
    for (const msg of messages) {
      if (msg.role === 'tool') continue;
      normalized.push(msg);
      if (msg.role === 'assistant' && msg.toolCalls && msg.toolCalls.length > 0) {
        for (const toolCall of msg.toolCalls) {
          if (toolCall.id == null || toolCall.id === '') continue;
          const toolMsg = deferredTools.get(toolCall.id);
          if (toolMsg) {
            deferredTools.delete(toolCall.id);
            normalized.push(toolMsg);
          } else {
            normalized.push({
              role: 'tool',
              toolCallId: toolCall.id,
              content: MISSING_TOOL_RESULT_PLACEHOLDER,
            });
            filled++;
          }
        }
      }
    }
    if (deferredTools.size > 0) {
      harnessLog('warn', `Dropping ${deferredTools.size} orphaned tool messages without a preceding assistant tool_calls`);
    }
    if (filled > 0) {
      harnessLog('warn', `Filled ${filled} missing tool output(s) before sending history to the LLM`);
    }
    return normalized;
  },
};

export function ensureContentPresent(messages: ChatMessage[] | null | undefined): void {
  MessageHistoryNormalizer.ensureContentPresent(messages);
}

function hasToolCallsOrToolMessages(messages: ChatMessage[]): boolean {
  for (const msg of messages) {
    if (msg.role === 'tool') return true;
    if (msg.role === 'assistant' && msg.toolCalls != null && msg.toolCalls.length > 0) return true;
  }
  return false;
}

function collectDeferredToolMessages(messages: Message[]): Map<string, Message> {
  const deferred = new Map<string, Message>();
  for (const msg of messages) {
    if (msg.role === 'TOOL' && msg.toolCallId != null) {
      deferred.set(msg.toolCallId, msg);
    }
  }
  return deferred;
}

function appendMatchingToolMessages(normalized: Message[], deferred: Map<string, Message>, ids: string[]): void {
  for (const id of ids) {
    const toolMsg = deferred.get(id);
    if (toolMsg) {
      deferred.delete(id);
      normalized.push(toolMsg);
    }
  }
}

function extractToolCallIds(toolCallsJson: string, parseToolCalls: (json: string) => ToolCall[]): string[] {
  const ids: string[] = [];
  try {
    const toolCalls = parseToolCalls(toolCallsJson);
    for (const tc of toolCalls) {
      if (tc.id != null) ids.push(tc.id);
    }
  } catch (e) {
    harnessLog('warn', `Failed to parse tool_calls while normalizing message history: ${(e as Error).message}`);
  }
  return ids;
}
